### DISCUSSION TITLE
Frame Seekable Timeline

### DISCUSSION THREAD

**Discussion by SubBass100** [aaronwmorris/indi-allsky#1091] [Ideas]
Indi-Allsky is one of my favorite projects, it's like an NVR for the sky. I love going back to the footage when I see something on social media saying "I saw something weird in the sky" and then finding it (usually starlink), fireworks, drones, airplanes, blimps, lightning etc. One thing I often find myself doing quite often is loading up the video and dragging the playhead forward and back on the video until I see the object in question then drag the play head back and try to pause on the best frame I want to save. 

As I understand there may be some trouble creating a frame by frame player with MP4 since it is compressed with keyframes etc, however, INDI-Allsky technically saves the high res frames for an amount of time right? It would be neat to have a tab similar to the "Loop" Tab combined with the "Images" tab which would allow you to play back the stills that you have, but with the ability to pause, and move frame by frame forward and backwards. Maybe section it off per day so the timeline would always be 24 hours at a time?

Binding the functions to the keyboard something like:

spacebar : pause 
Q : Play backwards (like loop)
W : Play forwards (like loop)
A : Move one frame backward
S : Move one frame forward
Z : Move 10 frames backward
X : Move 10 frames forward

Hopefully others would find this useful, I know I would. Thoughts?

**Comment by aaronwmorris** [comment]
By default, the original images are saved for 30 days, and this is configurable.

The Image viewer does some of this already, the bindings are currently for the arrow keys.  Left/right moves backward/forwards for each image during hour.  Up/Down increases/decreases the hour.  There is also a `Loop` option at the bottom that will animate for that point in time.  There is also some logic to increment/decrement the hour when you hit the first/last image of the hour.

I have also been considering the ability to create a short video based on a smaller time window to be able to share individual events.

I would have to be careful with loading a full days worth of images, since this would be an easy way to DoS your server.  It would not be able to be a public view.  The reason it is a DoS is below.

The real trouble with this is they way the images are loaded using the HTML5 Canvas element.  Loading/downloading individual images is an asynchronous event.  Meaning when the javascript is executed to display the image, there is no guarantee the image is fully downloaded by the browser.  This is the reason the Loop view can be jumpy/out-of-order when it first loads.  I have not found a way to wait for the image to be downloaded before trying to display it.

This also means that the browser will attempt to download all of the images in the loop nearly simultaneously, which easily saturates the network bandwidth.

**Reply by SubBass100** [reply]
Wow! I didn't realize that's how the "images" tab works, really cool about the key binds as well! I should have read the into it more. Totally understandable about the dos attack. I forgot that alot of people use this public facing. Also that makes a lot of sense about images being out of order. When I ran indi-allsky on a machine with weak resources and over wifi or vpn, I did notice that loop video was out of order. Since then I've installed on a vm with better resources and loops have been buttery smooth. Thanks again!

**Comment by aaronwmorris** [comment]
Merged #1093 to add WASD and VIM key bindings for image viewer.
